File: 프로그래머스/3/138475. 억억단을 외우자/억억단을 외우자.java
import java.util.*;

class Solution {
    public int[] solution(int e, int[] starts) {
        int[] count = new int[e+1];
        for (int i=1;i<=e;i++) {
            for (int j=1;j<=e/i;j++) {
                count[i*j]++;
            }
        }
        int maxN = e;
        int maxCount = count[e];
        count[e] = maxN;
        for (int i=e-1;i>0;i--) {
            if (count[i] >= maxCount) {
                maxCount = count[i];
                maxN = i;
            }
            count[i] = maxN;
        }
        ArrayList<Integer> answer = new ArrayList<>();
        for (int start: starts) {
            answer.add(count[start]);
        }
        return answer.stream().mapToInt(x -> x).toArray();
    }
}

File: 프로그래머스/3/138475. 억억단을 외우자/억억단을 외우자.py
def solution(e, starts):
    count = [0] * (e+1)
    for i in range(1, e+1):
        count[i] = get_count(i)

    max_n = e
    max_count = count[e]
    count[e] = max_n
    for i in range(e-1, 0, -1):
        c = count[i]
        if c >= max_count:
            max_count = c
            max_n = i
        count[i] = max_n

    answer = []
    for start in starts:
        answer.append(count[start])

    return answer

def get_count(n):
    root = int(n ** 0.5)
    result = 0
    for i in range(1, root+1):
        if n % i != 0: continue
        if i ** 2 == n:
            result += 1
        else:
            result += 2
    return result
    